tree-walk-xl.go

```diff
@@ -138,23 +140,37 @@ func (xl xlObjects) treeWalk(bucket, prefixDir, entryPrefixMatch, marker string,
 				// recursing into "four/"
 				markerArg = markerBase
 			}
-			*count--
 			prefixMatch := "" // Valid only for first level treeWalk and empty for subdirectories.
-			if !xl.treeWalk(bucket, pathJoin(prefixDir, entry), prefixMatch, markerArg, recursive, send, count, isLeaf) {
-				return false
+			if i == len(entries)-1 && stackDepth == 0 {
+				isEnd = true
 			}
+			stackDepth++
+			xl.treeWalk(bucket, pathJoin(prefixDir, entry), prefixMatch, markerArg, recursive, isLeaf, treeWalkCh, doneCh, stackDepth, isEnd)
+			stackDepth--
 			continue
 		}
-		*count--
-		if !send(treeWalkResult{entry: pathJoin(prefixDir, entry)}) {
-			return false
+		var isEOF bool
+		if stackDepth == 0 && i == len(entries)-1 {
+			isEOF = true
+		} else if i == len(entries)-1 && isEnd {
+			isEOF = true
 		}
```

Comment:
for the following directory tree structure, file.2 will not be listed as isEOF will be true when it sends file.1 down the channel

```
root/
├── a
└── b
    ├── c
    │   └── file.1
    └── d
        └── file.2
```